[AIFI] Add new object type AIFI

D052766 opened the pull request.

Markus1812 commented:
Hi, thank you for submitting your AFF. Before we start with the review, please address the abaplint issues. They are mostly regarding unknown ABAP types, as you can only use standard ABAP types or pre-defined types from our types interface: [zif_aff_types_v1.intf.abap](https://github.com/SAP/abap-file-formats/blob/486b1465bca61289af86d81cdd88af0fe5a8cf4f/file-formats/zif_aff_types_v1.intf.abap#L71).

For instance, `/aif/proxy_outbound` (char 30) can be replaced with `ty_object_name_30` from zif_aff_types_v1.

Please also double-check that all titles are in title-case and descriptions follow the sentence-case scheme as described in our [documentation](https://github.com/SAP/abap-file-formats/blob/main/docs/json.md#title).

```
"! <p class="shorttext">Integration Type</p>         <- title in title case
"! Integration type                                  <- description in sentence case
"! $required
integration_type TYPE ty_integration_type,
```

GuilhermeSaraiva96 reviewed (changes requested):
Commenting to change status to 'Changes Requested'

D052766 commented:
> Hi, thank you for submitting your AFF. Before we start with the review, please address the abaplint issues. They are mostly regarding unknown ABAP types, as you can only use standard ABAP types or pre-defined types from our types interface: [zif_aff_types_v1.intf.abap](https://github.com/SAP/abap-file-formats/blob/486b1465bca61289af86d81cdd88af0fe5a8cf4f/file-formats/zif_aff_types_v1.intf.abap#L71).
> 
> For instance, `/aif/proxy_outbound` (char 30) can be replaced with `ty_object_name_30` from zif_aff_types_v1.
> 
> Please also double-check that all titles are in title-case and descriptions follow the sentence-case scheme as described in our [documentation](https://github.com/SAP/abap-file-formats/blob/main/docs/json.md#title).
> 
> ```
> "! <p class="shorttext">Integration Type</p>         <- title in title case
> "! Integration type                                  <- description in sentence case
> "! $required
> integration_type TYPE ty_integration_type,
> ```

Hi, I have fixed now all issues that came up from the automated checks. Now that I cannot use the aif data types I think I need to add some enumerations. Also I think I need to double check whether all fields marked as required really need to be. Hope it is fine to have the pull request in place already so you can start the review!?

D052766 commented:
I have added some enumerations, moved sap and raw structure to General Information and removed several "required" annotations. I think there should be no more changes until your first feedback now. Any change of an estimation when that may be?

D052766 commented:
Related to feedback I got from Katharina, I´ve just committed another update on the structure.

Markus1812 commented:
Hi, thank you for updating the files. We'll perform a review tomorrow.

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Preprocessing</p>
+      "! Preprocessing
+      " /AIF/PRE_PROCESSING
+      preprocessing            TYPE abap_bool,
```
A suggestion from our side: For booleans we usually prefer the prefix `is_` to indicate a boolean in the json. And maybe for the `proxy_xml_transformation` a `uses_` prefix? 

But this is up to you.

D052766 replied:
I prefer to follow best practice and common format, so I will adjust it.

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Name of Select-Option/Parameter</p>
+      "! Name of select-option/parameter
+      " (type /aif/stexti)
+      selopt_name       TYPE c LENGTH 8,
```
```suggestion
      select_option_name       TYPE c LENGTH 8,
```
Maybe like this? Please don't use abbreviations.

D052766 replied:
ok

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Select-Option</p>
+      "! Is select-option?
+      " /AIF/FLD_IS_SELECT_OPT
+      is_selopt         TYPE abap_bool,
```
```suggestion
      is_select_option         TYPE abap_bool,
```

If you use the `is_` prefix here, maybe it would be better if all booleans are consistent with `is_`?

D052766 replied:
is_ and uses_ are now in place

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! Display as column
+      "! $default 'X'
+      " /AIF/IS_COL
+      column_display    TYPE abap_bool,
```
```suggestion
      is_column_display TYPE abap_bool,
```

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Message Index Table</p>
+      "! Message index table
+      " (type /aif/msg_tbl)
+      idx_table                TYPE zif_aff_types_v1=>ty_object_name_30,
```
```suggestion
      index_table                TYPE zif_aff_types_v1=>ty_object_name_30,
```
or maybe even `message_index_table`?

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Category Field Name</p>
+      "! Category field name
+      " (type /aif/alrt_catfn)
+      cat_fieldname            TYPE c LENGTH 32,
```
```suggestion
      category_field_name            TYPE c LENGTH 32,
```

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+  TYPES:
+    "! <p class="shorttext">Key Field</p>
+    "! Key field
+    BEGIN OF ty_key_field,
+      "! <p class="shorttext">General</p>
+      "! General key field settings
+      key_field_settings     TYPE ty_keyfield_settings,
+      "! <p class="shorttext">Key Field Determination by Qualifier</p>
+      "! Key field determination by qualifier
+      keyfield_qualifier     TYPE ty_keyfield_qualifier,
+      "! <p class="shorttext">Key Field Selection</p>
+      "! Key field selection type and settings
+      keyfield_selection     TYPE ty_keyfield_selection,
+      "! <p class="shorttext">Enable Key Field Rules</p>
+      "! Enable key field rules
+      key_field_rules_fields TYPE ty_key_field_rules_fields,
```
`key_field` or `keyfield`?

D052766 replied:
=> key_field

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Key Field Name</p>
+      "! Key field name
+      " (type /aif/key_fieldname_enh)
+      key_field_rule_field TYPE zif_aff_types_v1=>ty_object_name_30,
```
The name is a bit confusing with two `fields`. Do you have a better idea? Maybe `key_field_name` like in the title?

D052766 replied:
I agree, but it´s kind of tricky. It´s the name of a key field used in a key field rule. I propose 'rule_key_field_name'

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Key Fields</p>
+      "! Key fields
+      "! $required
+      o_key_fields         TYPE ty_o_key_fields,
```
Is there a reason for capsuling the table in `ty_o_key_fields`?

D052766 replied:
This is a leftover from trying out different things. I removed it.

Markus1812 reviewed (commented):
Hi, sorry for the delay. In general, the AFF looks really good, but we have a few questions:

1. Do you want to use prefixes like `is_` or `uses_` for booleans? We have marked some already and some already use prefixes. Maybe you can take a look again on that.
2. Do you want your comments containing the name of your domains/database tables in a public repository?

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">Sync Log</p>
+      "! Sync log
+      sync_log TYPE ty_integration_type VALUE 'Sync Log',
+      "! <p class="shorttext">pgPF</p>
+      "! pfPF
+      pg_pf    TYPE ty_integration_type VALUE 'pfPF',
```
A lot of typos on my end. It is always bgPF (Background Processing Framework) => I will correct that.

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">SOAP Class</p>
+      "! SOAP class
+      "! $required
+      " SOAP / web service class is stored in /AIF/T_FINF-PROXYCLASSNAMECL type /aif/proxy_outbound
```
I would like to keep these comments. They helped with the AFF creation and they will help later when implementing the workbench integration. Also only the names of the data elements do not expose much info. I will move the position.

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! Operation variant
+      "! $required
+      " bgPF (type /aif/bgp_variant_e)
+      operation_varaint    TYPE string,
```
Of course :-)

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,506 @@
+      "! <p class="shorttext">BgPF Settings</p>
+      "! BgPF settings
+      bgpf_settings            TYPE ty_bgpf_settings,
+      "! <p class="shorttext">Namespace</p>
+      "! Namespace
+      "! $required
+      " (type /aif/ns_obj_name)
+      namespace                TYPE c LENGTH 15,
```
Hm. The AIF Namespace is an attribute of AIFN. This is the AIFN object name.

D052766 commented:
I think all comments were answered and adjustments are done.

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      " (type /aif/alrt_catfn)
+      "! <p class="shorttext">Category Field Name</p>
+      "! Category field name
+      category_fieldname          TYPE c LENGTH 32,
```
```suggestion
      category_field_name         TYPE c LENGTH 32,
```
 
to match with `alert_field_name`

D052766 replied:
ok

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! <p class="shorttext">Key Field Selection</p>
+      "! Key field selection type and settings
+      key_field_selection  TYPE ty_key_field_selection,
+      "! <p class="shorttext">Enable Key Field Rules</p>
+      "! Enable key field rules
+      rule_key_field_names TYPE ty_rule_key_field_names,
```
Why do you have `Enable` in the title? Aren't these just `Key Field Rules`?

D052766 replied:
corrected

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! <p class="shorttext">Display in Message Monitoring</p>
+      "! Display in message monitoring
+      "! $default 'X'
+      is_message_monitor_show    TYPE abap_bool,
```
```suggestion
      is_shown_in_message_monitor    TYPE abap_bool,
```
maybe like this?

D052766 replied:
ok

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      " /AIF/IS_DEFAULT_FIELD
+      "! <p class="shorttext">Display in Message Monitoring by Default</p>
+      "! Display in message monitoring by default
+      is_message_monitor_default TYPE abap_bool,
```
```suggestion
      is_shown_by_default TYPE abap_bool,
```
`is_shown_in_message_monitor_by_default` would be nice but it's too long :(

Markus1812 reviewed (commented):
Thank you for adapting our suggestions. We have some more questions.

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! <p class="shorttext">Display as Column</p>
+      "! Display as column
+      "! $default 'X'
+      is_column_display  TYPE abap_bool,
```
We are a bit unsure again about `is_column_display`. Does it have the same meaning as `Display as Column`? What do you think? Maybe `display_as_column`?

D052766 replied:
This indicates whether the field is displayed as a column or not displayed at all. You cannot display it in an alternative fashion. I changed the description a little bit.

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! <p class="shorttext">Show Tree Node</p>
+      "! Show tree node in view 1 tree
+      "! $default 'X'
+      is_show_tree_node           TYPE abap_bool,
```
We have two verbs directly next to each other. Maybe `show_tree_node` or `is_tree_node_visible`?

D052766 replied:
Hm. I think show_tree_node contradicts your recommendation for boolean, right? I changed it to is_tree_node_visible .

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! <p class="shorttext">Display in Message Monitoring</p>
+      "! Display in message monitoring
+      "! $default 'X'
+      is_show_in_message_monitor TYPE abap_bool,
```
```suggestion
      is_shown_in_message_monitor TYPE abap_bool,
```

D052766 replied:
ok

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! <p class="shorttext">Key Fields</p>
+      "! Key fields
+      "! $required
+      key_fields_table     TYPE ty_key_fields,
```
```suggestion
      key_fields           TYPE ty_key_fields,
```

D052766 replied:
ok

Markus1812 reviewed (commented):
Thank you again for adapting your AFF. We have some small comments again.

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,496 @@
+      "! Icon tooltip
+      icon_tooltip                TYPE c LENGTH 40,
+      " (type /aif/alertfieldname)
+      "! <p class="shorttext">Alert Recipient Field Name</p>
+      "! Field name in alert recipient assignment table
+      alert_field_name            TYPE c LENGTH 45,
```
Hm. This is a little bit tricky. Alerts require a recipient. But the recipient is also used for determination of the available integrations in the monitor. So I would say everything is a recipient, but a recipient is required to activate alerts. I adjusted the text for the field in ty_recipient , so it´s clear that this is a more generic thing.

Markus1812 replied:
Ok, thanks 👍

Markus1812 commented on file-formats/aifi/examples/z_aff_example_aifi.aifi.json:
```diff
@@ -0,0 +1,153 @@
+      "operationVariant": "OP2023"
+    },
+    "namespace": "/AIF/",
+    "interfaceName": "IFTEST",
+    "interfaceVersion": "1",
```
One more question for clarification: Is this representing the name and version of the current object or a reference to another interface?

If this is representing the name and version of this current object, it would be redundant as the filename represents the object name and the (aff) version is already contained in the `formatVersion`.

D052766 replied:
This is AIF internal content which is stored in the object and the values are used by AIF framework only.

Markus1812 replied:
So if the object name would be `Z_AFF_EXAMPLE_AIFI` in the `TADIR`, this `interfaceName` would have some other value than `Z_AFF_EXAMPLE_AIFI`?

D052766 replied:
Correct

Markus1812 replied:
Ok, thanks

Markus1812 reviewed (commented):
In general, the AFF looks good to me, but I have one further question for clarification.

Markus1812 commented:
Thank you for answering all our questions. The AFF looks good to me, you can now continue and request the ux-review when you are ready for that.

D052766 commented:
I just removed the enumeration for the integration type. We have to supply this as custom value help from the backend. Please confirm the UX-REVIEW readiness is still given.

Markus1812 commented:
@D052766 Yes sure, thats okay for me. Is there are reason behind this change? Maybe this would be interesting for the ux colleagues as well. 

If your concern is the extensibility of the enum, our [wiki](https://github.com/SAP/abap-file-formats/blob/main/docs/json.md#compatible-file-format-changes) states that adding values to the enum is compatible without increasing the format version, as long as there is a default value (which is the case).

D047539 commented:
@Markus1812 We prefer to set and retrieve values in the backend, rather than defining them in the frontend.

D052766 commented:
After AIF internal discussions and review I had to do some more adjustments. I think we´re ready for UX review now, but I will wait for your comments.

Markus1812 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -139,72 +144,78 @@ INTERFACE zif_aff_aifi_v1
       "! Integration type
-      integration_type              TYPE ty_integration_type,
+      integration_type             TYPE c LENGTH 4,
+      "! <p class="shorttext">Integration Type</p>
```
Is this a `Integration Type Description`? Or is there a reason for leaving out `Description`? 

Same for `scenario_description`.

D052766 replied:
Yes. The description field was added, in order to display the text instead of the value (we don´t want an enumeration). But the field label should remain "Integration Type".

D052766 replied:
Same for the Scenario.

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -349,12 +360,12 @@ INTERFACE zif_aff_aifi_v1
-    "! Rule key field name
+    "! <p class="shorttext">Rule Key Field</p>
+    "! Rule key field
     BEGIN OF ty_rule_key_field_name,
```
ops. inconsistent change on my part. we have key fields and keyfield rules (rule belongs to key field) and again rule key fields (key fields belong to keyfile rule). I corrected it in will commit the changes today.

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -366,25 +377,40 @@ INTERFACE zif_aff_aifi_v1
-      key_field_selection  TYPE ty_key_field_selection,
+      key_field_selection TYPE ty_key_field_selection,
       "! <p class="shorttext">Key Field Rules</p>
       "! Key field rules
```
yes. respect you saw that ;-) I corrected it and will commit again today.

D052766 commented:
Key Field related wording has been reworked (I hope ;-))

Markus1812 commented:
Thanks for the update. Ready for UX review 👌

wurzka commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+    "! <p class="shorttext">Key Field</p>
+    "! Key field
+    BEGIN OF ty_key_field,
+      "! <p class="shorttext">General</p>
+      "! General key field settings
+      key_field_settings  TYPE ty_key_field_settings,
```
What about
```suggestion
      "! <p class="shorttext">Key Field Details</p>
      "! General key field settings
      ky_field_details  TYPE ty_key_field_settings,
```
? This would match to our usual recommendation with Plural (Key Fields) for Tables and Singular + Details for the Substructure (Key Field Details)

wurzka commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+    "! <p class="shorttext">Rule Key Field</p>
+    "! Rule key field
+    BEGIN OF ty_rule_key_field,
+      " (type /aif/key_fieldname_enh)
+      "! <p class="shorttext">Rule Key Field Name</p>
+      "! Rule Key field name
+      rule_key_field_name  TYPE zif_aff_types_v1=>ty_object_name_30,
```
Maybe 
```suggestion
      " (type /aif/key_fieldname_enh)
      "! <p class="shorttext">Name</p>
      "! Rule Key field name
      name  TYPE zif_aff_types_v1=>ty_object_name_30,
```
is enough here. Otherwise, you have in structure Key Field Rule the field Key Field Rule Name which is duplicate info.

D052766 replied:
Honestly I do not really get the point, but I adjusted it. Commit will follow.

wurzka commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+    "! <p class="shorttext">General</p>
+    "! General key field settings
+    BEGIN OF ty_key_field_settings,
+      " (type /aif/key_fieldname, field of index table)
+      "! <p class="shorttext">Key Field Name</p>
+      "! Key field name
+      "! $required
+      key_field_name     TYPE zif_aff_types_v1=>ty_object_name_30,
```
Also here: When the structure is already about Key Fields this
```suggestion
      " (type /aif/key_fieldname, field of index table)
      "! <p class="shorttext">Name</p>
      "! Key field name
      "! $required
      name     TYPE zif_aff_types_v1=>ty_object_name_30,
```
could be enough.

D052766 replied:
Ok. Now I understand. You´re saying the context of 'name' is given by the structure and doesn´t have to be given for a single field of it. Adjusted. Commit will follow.

wurzka commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+  TYPES ty_sap_raw_structure_type TYPE c LENGTH 1.
+  CONSTANTS:
+    BEGIN OF co_sap_raw_structure_type,
+      "! <p class="shorttext">Source Structure (raw for inbound, SAP for outbound)</p>
+      "! Source structure (raw for inbound, SAP for outbound)
+      raw TYPE ty_sap_raw_structure_type VALUE ' ',
+      "! <p class="shorttext">Destination Structure (SAP for inbound, raw for outbound)</p>
+      "! Destination structure (SAP for inbound, raw for outbound)
+      sap TYPE ty_sap_raw_structure_type VALUE 'S',
```
We strongly recommend having name of component and title in sync.

D052766 replied:
The title is what appears in the drop down. How would I make a component name from that? I think that´s not prudent here.

wurzka replied:
Maybe I don't understand this field correctly.
The field has currently two possible values: `Source Structure (raw for inbound, SAP for outbound)` and `Destination Structure (SAP for inbound, raw for outbound)`. Therefore I understood that raw is the technical value for Source Structure and sap for Destination structure. But checking your title again, it seems to be different?!

D052766 replied:
RAW and SAP structure can be source or destination, depending on the interface direction. Only thing I can imagine to simply this: Change Titles to Source Structure and Destination Structure. Leave away the hint in brackets. And then rename the components to source_structure and destination_structure . Would that be acceptable and comprehensible to you? I think can move the complicated detail here to the documentation.

wurzka replied:
For me, it would be less confusing. But I understood it, now. As expert on your domain, I would like to leave this decision up to you.

D052766 replied:
hm. in fact never felt comfortable with that state and adjusted it. pls double check from your end.

wurzka replied:
Thanks, looks good to me!

wurzka reviewed (changes requested):
Hi @D052766 ,
I went through the naming of the components and titles, please see my comments

D052766 commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+    BEGIN OF ty_general_information,
+      "! <p class="shorttext">Integration Type</p>
+      "! Integration type
+      integration_type             TYPE c LENGTH 4,
+      "! <p class="shorttext">Integration Type</p>
+      "! Integration type description
+      integration_type_description TYPE c LENGTH 40,
```
We do not want do declare the values on the UI/AFF, therefore enum is not desired. But I understand you proposal is to write the description to the INTEGRATION_TYPE instead of having an extra field. **I will check if we can do that and post another reply.**

D052766 replied:
Adjustment is done and works. Commit will follow combined in case further adjustment is required.

wurzka commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+    BEGIN OF ty_key_field,
+      "! <p class="shorttext">General</p>
+      "! General key field settings
+      key_field_settings  TYPE ty_key_field_settings,
```
Remove this structure and put its field in 'ty_key_field'

D052766 replied:
ok, understood.

wurzka commented on file-formats/aifi/type/zif_aff_aifi_v1.intf.abap:
```diff
@@ -0,0 +1,523 @@
+    END OF ty_key_field_rule.
+
+  TYPES:
+    "! <p class="shorttext">Key Field</p>
+    "! Key field
```
```suggestion
    "! <p class="shorttext">Key Field Details</p>
    "! Key field details
```

wurzka reviewed (changes requested):
Thanks for the adaption of my comments. UX check revealed one more issue, see my comment

wurzka reviewed (approved):
Thanks, looks good to me.
